dev/io.openliberty.org.apache.myfaces.4.0_fat/fat/src/io/openliberty/org/apache/myfaces40/fat/tests/WebSocketTests.java

```diff
@@ -0,0 +1,125 @@
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        ShrinkHelper.defaultDropinApp(server, WEB_SOCKET_TEST_APP_NAME + ".war",
+                                      "io.openliberty.org.apache.faces40.fat.websocket");
+
+        // Start the server and use the class name so we can find logs easily.
+        server.startServer(WebSocketTests.class.getSimpleName() + ".log");
+    }
+
+    @AfterClass
+    public static void tearDown() throws Exception {
+        // Stop the server
+        if (server != null && server.isStarted()) {
+            server.stopServer("SRVE0190E"); // SRVE0190E is due to ENABLE_WEBSOCKET_ENDPOINT being false for triggering onerror on a websocket.
+        }
+    }
+
+    @Before
+    public void setupPerTest() throws Exception {
+        server.setMarkToEndOfLog();
+    }
+
+    /**
+     * Test to ensure that the <f:websocket> onerror listener works properly.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testOnErrorWebsocket() throws Exception {
+        try (WebClient webClient = new WebClient()) {
+            webClient.getOptions().setThrowExceptionOnScriptError(false);
+
+            // Construct the URL for the test
+            String contextRoot = "WebSocket";
+            URL url = JSFUtils.createHttpUrl(server, contextRoot, "OnErrorWebSocketTest.jsf");
+
+            HtmlPage testOnErrorWebSocketPage = (HtmlPage) webClient.getPage(url);
+
+            // Log the page for debugging if necessary in the future.
+            Log.info(c, name.getMethodName(), testOnErrorWebSocketPage.asText());
+            Log.info(c, name.getMethodName(), testOnErrorWebSocketPage.asXml());
+
+            // Verify that the page contains the expected messages.
+            assertContains(testOnErrorWebSocketPage.asText(),
+                           "JSF 4.0 WebSocket - Test that onerror is invoked correctly.");
+
+            // Get the form that we are dealing with
+            HtmlForm form = testOnErrorWebSocketPage.getFormByName("form1");
+
+            // Get the button that opens the push connection
+            HtmlSubmitInput openButton = form.getInputByName("form1:openButton");
+
+            // Now click the open button. This should result in an error since 'jakarta.faces.ENABLE_WEBSOCKET_ENDPOINT' is set to 'false'
+            HtmlPage openPage = openButton.click();
+
+            assertTrue(JSFUtils.waitForPageResponse(openPage, "Called onerror listener"));
+            assertTrue(JSFUtils.waitForPageResponse(openPage, "Called onclose listener"));
+
+            //TODO: Should the onClose(@Observes @Closed WebsocketEvent event) method in WebsocketPushBean get invoked for onclose?
+//            String result = server.waitForStringInLogUsingMark("Channel myChannel was closed successfully!");
+//
+//            // Verify that the correct message is found in the logs
+//            assertNotNull("Message not found: 'Channel myChannel was closed successfully!'", result);
```

Comment:
We have an existing test (albeit commented out as of now) that tests the onOpen and onClose events. 

This check is probably not needed, unless I'm misunderstanding something.

Reply:
I see you mentioned this a slack comment 2 days ago. Let me review that first.

Reply:
So the onerror causes an onclose event to occur and the question is whether `onClose(@Observes @Closed WebsocketEvent event)` should be invoked?

Reply:
Right the question is should onClose() be invoked due to the onclose event triggered by the onerror.

Reply:
I don't think so since the websocket conn is never opened in the first place.
Logs only show: 
``
[2/10/23, 13:12:28:793 EST] 00000029 com.ibm.ws.webcontainer.extension                            W SRVE0190E: File not found: /jakarta.faces.push/myChannel
```